### PR TITLE
feat: introduce namespace config and new websocket transmission

### PR DESCRIPTION
**Pull request by lukeocodes**
Large refactor of internal code. New changes:

-   **Namespace Configuration**: Implementation of namespace options to manage configurations more effectively.
-   **WebSocket Transmission**: Introduction of new WebSocket logic to allow users to provide custom websocket transmitters, as well as reconnect the connection and provide API key factories to the websocket config (for client-side reconnections using short lived tokens).
-   **REST client transmission**: Introduction of new logic around `fetch` to allow users to provide a custom `fetch` interface, as well as expose more native fetch options to the client configuration.
-   **Client Abstractions**: Rebuilt clients to support namespace options.
-   **Config and Class Interfaces**: Rebuilt configuration and class interfaces for better modularity.
-   **Legacy Options Conversion**: Conversion of all legacy options to the new namespaced options.
-   **Live Client Integration**: Integration of live client with the new options.
-   **REST Client Integration**: Integration of REST clients with the new options.
-   **Documentation Updates**: Updated README and examples to reflect the new changes.

### CODE REVIEW

File: test/fetch.test.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  it("should not overwrite existing Authorization header", async () => {
+    const apiKey = "test-api-key";
+    const customFetch = async (input: RequestInfo | URL, init?: RequestInit | undefined) => {
+      const headers = new Headers(init?.headers);
+      expect(headers?.get("Authorization")).to.equal("existing-token");
+      return new Response();
+    };
+    const fetchWithAuthInstance = fetchWithAuth(apiKey, customFetch);
+    await fetchWithAuthInstance("https://example.com", {
+      headers: { Authorization: "existing-token" },
```
[the remark this answers is not in the document]
Comment (lukeocodes) [reply test/fetch.test.ts:25]: IT'S A TEST MORON 😂

**Comment by lukeocodes** [comment]
@jpvajda this is ready for review now. But it's so big, I recommend we all clone it and run the examples

File: README.md [hunk 2]
```diff
@@ -130,28 +148,84 @@ const deepgram = createClient(DEEPGRAM_API_KEY);
 
 # Scoped Configuration
 
-A new feature is scoped configuration. You'll be able to configure various aspects of the SDK from the initialization.
+The SDK supports scoped configurtion. You'll be able to configure various aspects of each namespace of the SDK from the initialization. Below outlines a flexible and customizable configuration system for the Deepgram SDK. Here’s how the namespace configuration works:
```
Comment (SandraRodgers) [inline README.md:151]: spelling mistake: configuration

File: examples/node-prerecorded/index.js [hunk 3]
```diff
@@ -4,31 +4,32 @@ const fs = require("fs");
...
 const transcribeFile = async () => {
   const deepgram = createClient(process.env.DEEPGRAM_API_KEY);
 
-  const { result, error } = await deepgram.listen.prerecorded.transcribeFile(
-    fs.readFileSync("./examples/nasa.mp4"),
-    {
-      model: "nova",
-    }
-  );
+  const file = fs.readFileSync("./examples/spacewalk.wav");
```
Comment (SandraRodgers) [inline examples/node-prerecorded/index.js:24]:
```suggestion
 const filePath = path.join(__dirname, '../spacewalk.wav');
const file = fs.readFileSync(filePath);
```

Comment (SandraRodgers) [reply examples/node-prerecorded/index.js:24]: If you choose to add that code, would also need to add the path module:
```js
const path = require('path');
```

**Review by SandraRodgers — DISMISSED** [review]

File: examples/node-prerecorded/index.js (hunk 3 above)
Comment (lukeocodes) [reply examples/node-prerecorded/index.js:24]: Thanks, I'll use this for local files then

**Review by jpvajda — APPROVED** [review]

File: examples/browser-prerecorded/index.html [hunk 4]
```diff
@@ -4,7 +4,7 @@
     <script src="../../dist/umd/deepgram.js"></script>
     <script>
       const { createClient } = deepgram;
-      const _deepgram = createClient("deepgram-api-key", {
+      const _deepgram = createClient("3d75575499a7908f7f01483469c8105897e37c1a", {
```
Comment (SandraRodgers) [inline examples/browser-prerecorded/index.html:7]: API Key

**Review by SandraRodgers — APPROVED** [review]

**Pull request merged**